### PR TITLE
mgr/cephadm: Adding --storage.tsdb.retention.size prometheus option

### PR DESCRIPTION
**Pull request by rkachach**
This PR adds support for option

Example of usage:
```
service_type: prometheus
placement:
  count: 1
spec:
  retention_time: "1y"
  retention_size: "512MB"
```
Once service is deployed:

```
[root@ceph-node-1 ~]# ps aux | grep retention.size
nobody     21411  0.0  0.0   1028     4 ?        Ss   12:55   0:00 /dev/init -- /bin/prometheus --config.file=/etc/prometheus/prometheus.yml --storage.tsdb.path=/prometheus --web.listen-address=:9095 --storage.tsdb.retention.time=1y --storage.tsdb.retention.size=512MB --web.external-url=http://ceph-node-1:9095

```

fixes: https://tracker.ceph.com/issues/57422

Signed-off-by: Redouane Kachach <rkachach@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: doc/cephadm/services/monitoring.rst [hunk 1]
```diff
@@ -344,10 +344,13 @@ Setting up Prometheus
 Setting Prometheus Retention Time
```
Comment (adk3798) [inline doc/cephadm/services/monitoring.rst:344]:
```suggestion
Setting Prometheus Retention Time and Size
```
I think we ought to mention both in the section title

File: doc/cephadm/services/monitoring.rst [hunk 2]
```diff
@@ -344,10 +344,13 @@ Setting up Prometheus
...
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Cephadm provides the option to set the Prometheus TDSB retention time using
-a ``retention_time`` field in the Prometheus service spec. The value defaults
-to 15 days (15d). If you would like a different value, such as 1 year (1y) you
-can apply a service spec similar to:
+Cephadm provides the option to set the Prometheus TDSB retention time specifying
+``retention_time`` and/or ``retention_size`` fields in the Prometheus service spec.
```
Comment (adk3798) [inline doc/cephadm/services/monitoring.rst:348]:
```suggestion
Cephadm provides the option to set the Prometheus TDSB retention time and/or size by specifying
the ``retention_time`` and``retention_size`` fields respectively in the Prometheus service spec.
```
Personal opinion on wording. Feel free to just mark this one resolved if you disagree.

File: doc/cephadm/services/monitoring.rst [hunk 3]
```diff
@@ -344,10 +344,13 @@ Setting up Prometheus
...
+The retention time value defaults to 15 days (15d). User can set different value/unit where
```
Comment (adk3798) [inline doc/cephadm/services/monitoring.rst:349]:
```suggestion
The retention time value defaults to 15 days (15d). Users can set a different value/unit where
```

File: src/pybind/mgr/cephadm/module.py [hunk 4]
```diff
@@ -2533,6 +2533,19 @@ def _apply(self, spec: GenericSpec) -> str:
             # should only refresh if a change has been detected
             self._trigger_preview_refresh(specs=[cast(DriveGroupSpec, spec)])
 
+        if spec.service_type == 'prometheus':
+            spec = cast(PrometheusSpec, spec)
+            if spec.retention_time:
+                valid_units = ['y', 'w', 'd', 'h', 'm', 's']
+                m = re.search(rf"^(\d+)({'|'.join(valid_units)})$", spec.retention_time)
```
Comment (adk3798) [inline src/pybind/mgr/cephadm/module.py:2540]: can we unit test this function? I think we ought to have something small to make sure these regex we're using work as intended.

**Review by adk3798 — COMMENTED** [review]
Generally LGTM, just some minor comments and a test request.

File: doc/cephadm/services/monitoring.rst (hunk 2 above)
Comment (anthonyeleven) [reply doc/cephadm/services/monitoring.rst:348]: I'd suggest something more concise like

 "Cephadm can configure Prometheus TSDB retention by specifying ``retention_time`` and``retention_size`` values in the Prometheus service spec."

Comment (adk3798) [reply doc/cephadm/services/monitoring.rst:348]: > I'd suggest something more concise like
> 
> "Cephadm can configure Prometheus TSDB retention by specifying `retention_time` and`retention_size` values in the Prometheus service spec."

I like that better as well. @rkachach if you're going to change I'd say go with @anthonyeleven suggestion over mine.

File: src/pybind/mgr/cephadm/module.py (hunk 4 above)
Comment (rkachach) [reply src/pybind/mgr/cephadm/module.py:2540]: Added extended UT for retention time and retention size

File: doc/cephadm/services/monitoring.rst (hunk 3 above)
Comment (rkachach) [reply doc/cephadm/services/monitoring.rst]: fixed

File: doc/cephadm/services/monitoring.rst (hunk 2 above)
Comment (rkachach) [reply doc/cephadm/services/monitoring.rst]: fixed

File: doc/cephadm/services/monitoring.rst (hunk 1 above)
Comment (rkachach) [reply doc/cephadm/services/monitoring.rst]: fixed

**Comment by adk3798** [comment]
jenkins test make check

**Review by anthonyeleven — APPROVED** [review]

**Comment by rkachach** [comment]
jenkins test make check

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2022-09-12_11:00:28-orch:cephadm-wip-adk-testing-2022-09-11-1805-distro-default-smithi/

failures tracked by https://tracker.ceph.com/issues/57287 and unrelated to any PRs in this run

**Pull request merged**